METRICS.md

```diff
@@ -0,0 +1,282 @@
+)
+
+SELECT
+CASE WHEN daily THEN "Daily" 
+     WHEN weekly THEN "Weekly" 
+      # If at least one per month, then Monthly
+     WHEN PERCENTILE_CONT(monthly_deploys, 0.5) OVER () >= 1 THEN  "Monthly" 
+     ELSE "Yearly"
+     END as deployment_frequency
+FROM (
+  SELECT
+  # If the median number of days per week is more than 3, then Daily
+  PERCENTILE_CONT(days_deployed, 0.5) OVER() >= 3 AS daily,
+  # If most weeks have a deployment, then Weekly
+  PERCENTILE_CONT(week_deployed, 0.5) OVER() >= 1 AS weekly,
+
+  # Count the number of deployments per month.  
+  # Cannot mix aggregate and analytic functions, so calculate the median in the outer select statement
+  SUM(week_deployed) OVER(partition by TIMESTAMP_TRUNC(week, MONTH)) monthly_deploys
+  FROM(
+      SELECT
+      TIMESTAMP_TRUNC(last_three_months.day, WEEK) as week,
+      MAX(if(deployments.day is not null, 1, 0)) as week_deployed,
+      COUNT(distinct deployments.day) as days_deployed
+      FROM last_three_months
+      LEFT JOIN(
+        SELECT
+        TIMESTAMP_TRUNC(time_created, DAY) AS day,
+        deploy_id
+        FROM four_keys.deployments) deployments ON deployments.day = last_three_months.day
+      GROUP BY week)
+ )
+LIMIT 1;
+```
+
+## Lead Time for Changes ##
+
+**Definition**: The median amount of time for a commit to be deployed into production.
+
+### Daily Median Lead Time ###
+![Image of chart from the Four Keys dashboard, showing the daily median lead time.](images/daily_lead_time.png)
+
+
+If we have a list of all changes for every deployment, it is simple to calculate the lead time for each change to deployment.
+
+```sql
+SELECT
+ d.deploy_id,
+ TIMESTAMP_TRUNC(d.time_created, DAY) as day,
+ ##### Time to Change
+ TIMESTAMP_DIFF(d.time_created, c.time_created, MINUTE) time_to_change_minutes
+FROM four_keys.deployments d, d.changes
+LEFT JOIN four_keys.changes c ON changes = c.change_id;
+```
+
+From this base, we want to extract the daily median lead time to change. 
+
+```sql
+SElECT
+day,
+PERCENTILE_CONT(
+  # Ignore automated changes
+  IF(time_to_change_minutes > 0,time_to_change_minutes, NULL), 
+  0.5) # Median
+  OVER (partition by day) median_time_to_change
+FROM
+(SELECT
+ d.deploy_id,
+ TIMESTAMP_TRUNC(d.time_created, DAY) as day,
+ # Time to Change
+ TIMESTAMP_DIFF(d.time_created, c.time_created, MINUTE) time_to_change_minutes
+ FROM four_keys.deployments d, d.changes
+ LEFT JOIN four_keys.changes c ON changes = c.change_id
+ )
+GROUP BY day, time_to_change_minutes;
+```
+
+It is up for debate whether or not we want to ignore the automated changes.  The rationale here is that when we merge a Pull Request it creates a Push event in the main branch.  This Push event is not its own distinct change, but rather a link in the workflow.  If we trigger a deployment off of this push event, this artificially skews the metrics and does not give us a clear picture of developer velocity. 
```

Comment:
It took me a minute to figure out which choice we actually made vis-a-vis including automated changes. Of course, there's a comment right up above, but for illiterates like me, I suggest changing this to something like "Automated changes are excluded from this metric. (This is a subject up for debate. The rationale is...)"

Additionally, it'd be great to link to an article or discussion that goes into further depth on the debate. If such a thing exists.

Reply:
Afraid, I don't have any links to share here, but I updated the language to make it clear what we were doing.